texk/lcdf-typetools/otfinfo/otfinfo.1
'\"t
.ds V 2.94
.de M
.BR "\\$1" "(\\$2)\\$3"
..
.de Sp
.if n .sp
.if t .sp 0.4
..
.TH OTFINFO 1 "LCDF Typetools" "Version \*V"
.SH NAME
otfinfo \- report information about OpenType fonts
.SH SYNOPSIS
.B otfinfo
\%[\fB-sfzpvag\fR]
\%[\fIFILE...\fR]
'
.SH DESCRIPTION
.BR Otfinfo
reports information about the named OpenType font
.IR FILE s
(or standard input, if no 
.IR FILE s
are given). Results are printed to standard output. Each line is prefaced
with the relevant
.I FILE
if more than one
.I FILE
was supplied.
'
.SH OPTIONS
With long options, you need type only as many characters as will make the
option unique.
.SS Query options
.PD 0
.PD 0
.TP 5
.BR \-s ", " \-\-scripts
Print supported scripts and language systems. The scripts are printed one
per line, with human-readable descriptions; for example:
.nf
  cyrl            Cyrillic
  grek            Greek
  latn            Latin
  latn.TUR        Latin/Turkish
.fi
'
.Sp
.TP 5
.BR \-f ", " \-\-features
Print GSUB and GPOS features supported by the selected script (see the
.B \-\-script
option below). The scripts are printed one per line, with human-readable
descriptions; for example:
.nf
  aalt    Access All Alternates
  c2sc    Small Capitals From Capitals
  case    Case-Sensitive Forms
  cpsp    Capital Spacing
  \&...
  zero    Slashed Zero
.fi
'
.Sp
.TP 5
.BR \-z ", " \-\-optical\-size
Print optical size information. For example:
.nf
  design size 11 pt, size range (8.4 pt, 13 pt], subfamily ID 11, subfamily name Semibold Italic
.fi
'
.Sp
.TP 5
.BR \-p ", " \-\-postscript\-name
Print each font's PostScript name. For example:
.nf
  MinionPro-SemiboldItCapt
.fi
'
.Sp
.TP 5
.BR \-a ", " \-\-family
Print each font's family name. For example:
.nf
  Minion Pro
.fi
'
.Sp
.TP 5
.BR \-v ", " \-\-font\-version
Print font version information, if available. For example:
.nf
  OTF 1.013;PS 001.000;Core 1.0.27;makeotf.lib(1.11)
.fi
'
.Sp
.TP 5
.BR \-i ", " \-\-info
Print each font's name, version, designer, vendor, copyright, and license
information. For example:
.nf
  Family:              Minion Pro
  Subfamily:           Regular
  Full name:           Minion Pro
  PostScript name:     MinionPro-Regular
  Version:             OTF 1.011;PS 001.000;Core 1.0.27;makeotf.lib1.3.1
  Unique ID:           1.011;ADBE;MinionPro-Regular
  Designer:            Robert Slimbach
  Vendor URL:          http://www.adobe.com/type/
  Trademark:           Minion is either a registered trademark or a trademark of Adobe Systems Incorporated in the United States and/or other countries.
  Copyright:           \(co 2000 Adobe Systems Incorporated. All Rights Reserved. U.S. Patent Des. 337,604. Other patents pending.
  License URL:         http://www.adobe.com/type/legal.html
.fi
'
.Sp
.TP 5
.BR \-g ", " \-\-glyphs
Print the name of every glyph in each font, one per line. For example:
.nf
  \&.notdef
  space
  exclam
  \&...
  ncommaaccent.end
  lje.alt
.fi
Works for PostScript-flavored OpenType fonts only.
'
.Sp
.TP 5
.BR \-t ", " \-\-tables
Print the size and name of every OpenType table in the font. For example:
.nf
       52 BASE
    87723 CFF
     4940 DSIG
    21366 GPOS
  \&...
.fi
'
.Sp
.TP 5
.BR \-T " \fItable\fR, " \-\-dump\-table= \fItable\fR
Print the contents of the font's OpenType table \fItable\fR.
'
.PD
'
'
.SS Miscellaneous options
'
.PD 0
.TP 5
.BI \-\-script= "script\fR[.\fIlang\fR]"
Select the script system
.I script
and language system
.IR lang
used to look up features by
.BR \-\-features .
Examples include "latn" (Latin script), "grek" (Greek script), and "yi.YIC"
(Yi script with classic characters). If
.I lang
is not specified, otfinfo will use the default language system for that
script. Defaults to "latn".
'
.Sp
.TP 5
.BR \-V ", " \-\-verbose
Write progress messages to standard error.
'
.Sp
.TP 5
.BR \-q ", " \-\-quiet
Do not generate any error messages.
'
.Sp
.TP 5
.BR \-h ", " \-\-help
Print usage information and exit.
'
.Sp
.TP 5
.BR \-\-version
Print the version number and some short non-warranty information and exit.
.PD
'
.SH "SEE ALSO"
.LP
.M otftotfm 1
.LP
.IR "OpenType Specification" ,
Version 1.4
'
.SH AUTHOR
Eddie Kohler (ekohler@gmail.com)


ChangeLog
2012-08-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Running texlinks in texk/texlive/linked_scripts.

2012-07-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build: Add TL_COMPILER_GFLAGS also to OBJCXXFLAGS.

2012-07-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-06-15.
	* version.ac: Switch 2012 => 2013/dev.

2012-06-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-06-15.

2012-05-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Test for GNU make before creating Makefile,
	required for FreeBSD make and perhaps others.

2012-05-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README: Update the list of prerequisites.
	* README.config: Mention configure options for biber.

2012-04-12  Karl Berry  <karl@tug.org>

	* Build (TL_COMPILER_GFLAGS): new envvar, used to override -g
	for all three compilers, with Build -g.

2012-04-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-04-03.

2012-03-31  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_LARGEFILE): switch default to --enable-largefile.
	More info in README.config.

2012-03-12  Karl Berry  <karl@tug.org>

	* README.config: 32-bit rather than 32Bit

2012-03-11  Karl Berry  <karl@tug.org>

	* README.config: text tweaks for poppler/libxpdf/luatex/etc.

2012-01-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Report `$enable_largefile'.

2012-01-13  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-01-13.

2011-12-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Replace DISTCHECK_CONFIGURE_FLAGS by
	AM_DISTCHECK_CONFIGURE_FLAGS (for automake-1.11.2).

2011-12-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Use TL specific installation directories only
	for $enable_native_texlive_build=yes (also datarootdir).
	Display the installation directories.

	* README.config, README.distro: Adapted.

2011-10-05  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-10-05.

2011-08-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-08-10.

2011-07-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am [EXTRA_DIST]: Add m4/ChangeLog.

2011-07-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2011 => 2012/dev.

2011-06-21  Karl Berry  <karl@tug.org>

	* Build: report final exit status.

2011-04-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-06-15.

2011-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am, configure.ac: Conditional 'cross_compiling' =>
	'cross' as in libs/icu, texk/dvipng, and texk/web2c.

	* version.ac: Switch 2011/dev => 2011.

2011-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-04-12.

2011-04-09  Karl Berry  <karl@tug.org>

	* Build (MACOSX_DEPLOYMENT_TARGET): no longer needed per koch.

2011-03-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-03-03.

2011-02-16  Karl Berry  <karl@tug.org>

	* Build: include uname -a in log, mention log if failure.

2011-01-19  Karl Berry  <karl@tug.org>

	* Build: don't try to count binaries if no bindir.	

2010-10-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac (new): Define the current TeX Live tarball version.
	* version.ac: Switch to 2011/dev.

2010-07-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (new): Define the current TeX Live version.

2010-05-18  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_BANNER): switch ident to /pretest.

2010-03-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build (TL_CONF_THREADS): Removed, now in libs/icu/configure.

2010-03-05  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (triptrap): Run intermediate triptrap target in
	 texk/, texk/web2c/Makefile might not yet exist.

2010-03-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (triptrap): New target to run TRIP and TRAP tests.
	Do not automatically run 'make check' when cross compiling.

2010-03-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am, configure.ac: Do not automatically run tests as
	part of 'make world' when cross compiling.

2010-02-22  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_BANNER): 2010/dev.

2009-12-24  Karl Berry  <karl@tug.org>

	* Makefile.am (world): also run make check.
	(check_target): new variable.
	(skip): new target, to do nothing.

	* Build (TL_CONF_BANNER): add /dev.

2009-12-22  Karl Berry  <karl@tug.org>

	* build-aux/ylwrap: update from current automake (no real changes).

2009-11-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Disable cross compilation of xindy.
	* configure.ac: Report absolute build and source dir; reject
	paths with embedded spaces.

2009-09-17  Karl Berry  <karl@tug.org>

	* Build: show Build command itself in build.log.

2009-09-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Report level of compiler warnings used.

2009-08-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (dist*check_listfiles): Stricter checking for
	'make distcheck': check for leftover files and symlinks.
	This really ought to be the Automake default.

2009-06-18  Karl Berry  <karl@tug.org>

	* Build: eradicate -Wdecl stuff, peb has implemented something
	more general.


texk/lcdf-typetools/otfinfo/otfinfo.cc
/* otfinfo.cc -- driver for reporting information about OpenType fonts
 *
 * Copyright (c) 2003-2011 Eddie Kohler
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version. This program is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif
#include <efont/psres.hh>
#include <efont/otfgsub.hh>
#include <efont/otfgpos.hh>
#include <efont/otfname.hh>
#include <efont/otfos2.hh>
#include <efont/cff.hh>
#include <lcdf/clp.h>
#include <lcdf/error.hh>
#include <lcdf/straccum.hh>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <stdarg.h>
#include <ctype.h>
#include <errno.h>
#include <algorithm>
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#if defined(_MSDOS) || defined(_WIN32)
# include <fcntl.h>
# include <io.h>
#endif

using namespace Efont;

#define VERSION_OPT		301
#define HELP_OPT		302
#define QUIET_OPT		303
#define VERBOSE_OPT		304
#define SCRIPT_OPT		305

#define QUERY_SCRIPTS_OPT	320
#define QUERY_FEATURES_OPT	321
#define QUERY_OPTICAL_SIZE_OPT	322
#define QUERY_POSTSCRIPT_NAME_OPT 323
#define QUERY_GLYPHS_OPT	324
#define QUERY_FVERSION_OPT	325
#define TABLES_OPT		326
#define QUERY_FAMILY_OPT	327
#define INFO_OPT		328
#define DUMP_TABLE_OPT		329

const Clp_Option options[] = {

    { "script", 0, SCRIPT_OPT, Clp_ValString, 0 },
    { "quiet", 'q', QUIET_OPT, 0, Clp_Negate },
    { "verbose", 'V', VERBOSE_OPT, 0, Clp_Negate },
    { "features", 'f', QUERY_FEATURES_OPT, 0, 0 },
    { "scripts", 's', QUERY_SCRIPTS_OPT, 0, 0 },
    { "optical-size", 'z', QUERY_OPTICAL_SIZE_OPT, 0, 0 },
    { "postscript-name", 'p', QUERY_POSTSCRIPT_NAME_OPT, 0, 0 },
    { "family", 'a', QUERY_FAMILY_OPT, 0, 0 },
    { "font-version", 'v', QUERY_FVERSION_OPT, 0, 0 },
    { "info", 'i', INFO_OPT, 0, 0 },
    { "glyphs", 'g', QUERY_GLYPHS_OPT, 0, 0 },
    { "tables", 't', TABLES_OPT, 0, 0 },
    { "dump-table", 'T', DUMP_TABLE_OPT, Clp_ValString, 0 },
    { "help", 'h', HELP_OPT, 0, 0 },
    { "version", 0, VERSION_OPT, 0, 0 },

};


static const char *program_name;

static Efont::OpenType::Tag script, langsys;

bool verbose = false;
bool quiet = false;


void
usage_error(ErrorHandler *errh, const char *error_message, ...)
{
    va_list val;
    va_start(val, error_message);
    if (!error_message)
	errh->message("Usage: %s [OPTION]... FONT", program_name);
    else
	errh->vxmessage(ErrorHandler::e_error, error_message, val);
    errh->message("Type %s --help for more information.", program_name);
    exit(1);
}

void
usage()
{
    FileErrorHandler uerrh(stdout);
    uerrh.message("\
%<Otfinfo%> reports information about an OpenType font to standard output.\n\
Options specify what information to print.\n\
\n\
Usage: %s [-sfzpg | OPTIONS] [OTFFILES...]\n\n",
	   program_name);
    uerrh.message("\
Query options:\n\
  -s, --scripts                Report font%,s supported scripts.\n\
  -f, --features               Report font%,s GSUB/GPOS features.\n\
  -z, --optical-size           Report font%,s optical size information.\n\
  -p, --postscript-name        Report font%,s PostScript name.\n\
  -a, --family                 Report font%,s family name.\n\
  -v, --font-version           Report font%,s version information.\n\
  -i, --info                   Report font%,s names and designer/vendor info.\n\
  -g, --glyphs                 Report font%,s glyph names.\n\
  -t, --tables                 Report font%,s OpenType tables.\n\
  -T, --dump-table NAME        Output font%,s %<NAME%> table.\n\
\n\
Other options:\n\
      --script=SCRIPT[.LANG]   Set script used for --features [latn].\n\
  -V, --verbose                Print progress information to standard error.\n\
  -h, --help                   Print this message and exit.\n\
  -q, --quiet                  Do not generate any error messages.\n\
      --version                Print version number and exit.\n\
\n\
Report bugs to <ekohler@gmail.com>.\n");
}

String
read_file(String filename, ErrorHandler *errh, bool warning = false)
{
    FILE *f;
    int f_errno = 0;
    if (!filename || filename == "-") {
	filename = "<stdin>";
	f = stdin;
#if defined(_MSDOS) || defined(_WIN32)
	// Set the file mode to binary
	_setmode(_fileno(f), _O_BINARY);
#endif
    } else {
	f = fopen(filename.c_str(), "rb");
	f_errno = errno;
    }

    String error_anno = (warning ? errh->e_warning : errh->e_error) + ErrorHandler::make_landmark_anno(filename);
    if (!f) {
	errh->xmessage(error_anno, strerror(f_errno));
	return String();
    }

    StringAccum sa;
    int amt;
    do {
	if (char *x = sa.reserve(8192)) {
	    amt = fread(x, 1, 8192, f);
	    sa.adjust_length(amt);
	} else
	    amt = 0;
    } while (amt != 0);
    if (!feof(f) || ferror(f))
	errh->xmessage(error_anno, strerror(errno));
    if (f != stdin)
	fclose(f);
    return sa.take_string();
}

String
printable_filename(const String &s)
{
    if (!s || s == "-")
	return String::make_stable("<stdin>");
    else
	return s;
}

static void
collect_script_descriptions(const OpenType::ScriptList &script_list, Vector<String> &output, ErrorHandler *errh)
{
    Vector<OpenType::Tag> script, langsys;
    script_list.language_systems(script, langsys, errh);
    for (int i = 0; i < script.size(); i++) {
	String what = script[i].text();
	const char *s = script[i].script_description();
	String where = (s ? s : "<unknown script>");
	if (!langsys[i].null()) {
	    what += String(".") + langsys[i].text();
	    s = langsys[i].language_description();
	    where += String("/") + (s ? s : "<unknown language>");
	}
	if (what.length() < 8)
	    output.push_back(what + String("\t\t") + where);
	else
	    output.push_back(what + String("\t") + where);
    }
}

static void
do_query_scripts(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    Vector<String> results;
    if (String gsub_table = otf.table("GSUB")) {
	OpenType::Gsub gsub(gsub_table, &otf, errh);
	collect_script_descriptions(gsub.script_list(), results, errh);
    }
    if (String gpos_table = otf.table("GPOS")) {
	OpenType::Gpos gpos(gpos_table, errh);
	collect_script_descriptions(gpos.script_list(), results, errh);
    }

    if (results.size()) {
	std::sort(results.begin(), results.end());
	String *unique_result = std::unique(results.begin(), results.end());
	for (String *sp = results.begin(); sp < unique_result; sp++)
	    result_errh->message("%s", sp->c_str());
    }
}

static void
collect_feature_descriptions(const OpenType::ScriptList &script_list, const OpenType::FeatureList &feature_list, Vector<String> &output, ErrorHandler *errh)
{
    int required_fid;
    Vector<int> fids;
    // collect features applying to this script
    script_list.features(script, langsys, required_fid, fids, errh);
    for (int i = -1; i < fids.size(); i++) {
	int fid = (i < 0 ? required_fid : fids[i]);
	if (fid >= 0) {
	    OpenType::Tag tag = feature_list.tag(fid);
	    const char *s = tag.feature_description();
	    output.push_back(tag.text() + String("\t") + (s ? s : "<unknown feature>"));
	}
    }
}

static void
do_query_features(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    Vector<String> results;
    if (String gsub_table = otf.table("GSUB")) {
	OpenType::Gsub gsub(gsub_table, &otf, errh);
	collect_feature_descriptions(gsub.script_list(), gsub.feature_list(), results, errh);
    }
    if (String gpos_table = otf.table("GPOS")) {
	OpenType::Gpos gpos(gpos_table, errh);
	collect_feature_descriptions(gpos.script_list(), gpos.feature_list(), results, errh);
    }

    if (results.size()) {
	std::sort(results.begin(), results.end());
	String *unique_result = std::unique(results.begin(), results.end());
	for (String *sp = results.begin(); sp < unique_result; sp++)
	    result_errh->message("%s", sp->c_str());
    }
}

static void
do_query_optical_size(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    int before_nerrors = errh->nerrors();
    try {
	String gpos_table = otf.table("GPOS");
	if (!gpos_table)
	    throw OpenType::Error();

	OpenType::Gpos gpos(gpos_table, errh);
	OpenType::Name name(otf.table("name"), errh);

	// extract 'size' feature
	int required_fid;
	Vector<int> fids;
	gpos.script_list().features(script, langsys, required_fid, fids, errh);

	int size_fid = gpos.feature_list().find(OpenType::Tag("size"), fids);
	if (size_fid < 0)
	    throw OpenType::Error();

	// old Adobe fonts implement an old, incorrect idea
	// of what the FeatureParams offset means.
	OpenType::Data size_data = gpos.feature_list().size_params(size_fid, name, errh);
	if (!size_data.length())
	    throw OpenType::Error();

	StringAccum sa;
	sa << "design size " << (size_data.u16(0) / 10.) << " pt";
	if (size_data.u16(2) != 0) {
	    sa << ", size range (" << (size_data.u16(6) / 10.) << " pt, "
	       << (size_data.u16(8) / 10.) << " pt], "
	       << "subfamily ID " << size_data.u16(2);
	    if (String n = name.english_name(size_data.u16(4)))
		sa << ", subfamily name " << n;
	}

	result_errh->message("%s", sa.c_str());

    } catch (OpenType::Error) {
	if (errh->nerrors() == before_nerrors)
	    result_errh->message("no optical size information");
    }
}

static void
do_query_family_name(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    int before_nerrors = errh->nerrors();
    String family_name = "no family name information";

    if (String name_table = otf.table("name")) {
	OpenType::Name name(name_table, errh);
	if (name.ok())
	    family_name = name.english_name(OpenType::Name::N_FAMILY);
    }

    if (errh->nerrors() == before_nerrors)
	result_errh->message("%s", family_name.c_str());
}

static void
do_query_postscript_name(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    int before_nerrors = errh->nerrors();
    String postscript_name = "no PostScript name information";

    if (String name_table = otf.table("name")) {
	OpenType::Name name(name_table, errh);
	if (name.ok())
	    postscript_name = name.english_name(OpenType::Name::N_POSTSCRIPT);
    }

    if (errh->nerrors() == before_nerrors)
	result_errh->message("%s", postscript_name.c_str());
}

static void
do_query_font_version(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    int before_nerrors = errh->nerrors();
    String version = "no version information";

    if (String name_table = otf.table("name")) {
	OpenType::Name name(name_table, errh);
	if (name.ok())
	    version = name.english_name(OpenType::Name::N_VERSION);
    }

    if (errh->nerrors() == before_nerrors)
	result_errh->message("%s", version.c_str());
}

static void
do_info(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    int before_nerrors = errh->nerrors();
    StringAccum sa;

    if (String name_table = otf.table("name")) {
	OpenType::Name name(name_table, errh);
	if (name.ok()) {
	    if (String s = name.english_name(OpenType::Name::N_FAMILY))
		sa << "Family:              " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_SUBFAMILY))
		sa << "Subfamily:           " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_FULLNAME))
		sa << "Full name:           " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_POSTSCRIPT))
		sa << "PostScript name:     " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_POSTSCRIPT_CID))
		sa << "PostScript CID name: " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_PREF_FAMILY))
		sa << "Preferred family:    " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_PREF_SUBFAMILY))
		sa << "Preferred subfamily: " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_MAC_COMPAT_FULLNAME))
		sa << "Mac font menu name:  " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_VERSION))
		sa << "Version:             " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_UNIQUEID))
		sa << "Unique ID:           " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_DESCRIPTION))
		sa << "Description:         " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_DESIGNER))
		sa << "Designer:            " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_DESIGNER_URL))
		sa << "Designer URL:        " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_MANUFACTURER))
		sa << "Manufacturer:        " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_VENDOR_URL))
		sa << "Vendor URL:          " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_TRADEMARK))
		sa << "Trademark:           " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_COPYRIGHT))
		sa << "Copyright:           " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_LICENSE_URL))
		sa << "License URL:         " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_LICENSE_DESCRIPTION))
		sa << "License Description: " << s << "\n";
	    if (String s = name.english_name(OpenType::Name::N_SAMPLE_TEXT))
		sa << "Sample text:         " << s << "\n";
	}
    }

    if (String os2_table = otf.table("OS/2")) {
	OpenType::Os2 os2(os2_table, errh);
	if (os2.ok()) {
	    if (String s = os2.vendor_id()) {
		while (s.length() && (s.back() == ' ' || s.back() == 0))
		    s = s.substring(s.begin(), s.end() - 1);
		if (s)
		    sa << "Vendor ID:           " << s << "\n";
	    }
	}
    }

    if (sa || errh->nerrors() == before_nerrors)
	result_errh->message("%s", (sa ? sa.c_str() : "no information"));
}

static void
do_query_glyphs(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    int before_nerrors = errh->nerrors();
    try {
	// get font
	Cff cff(otf.table("CFF"), errh);
	if (!cff.ok())
	    throw OpenType::Error();

	Cff::FontParent *fp = cff.font(PermString(), errh);
	if (!fp || !fp->ok())
	    throw OpenType::Error();
	Cff::Font *font = dynamic_cast<Cff::Font *>(fp);
	if (!font) {
	    errh->error("CID-keyed fonts not supported");
	    throw OpenType::Error();
	}

	// save glyph names
	Vector<PermString> glyph_names;
	font->glyph_names(glyph_names);
	for (PermString* s = glyph_names.begin(); s < glyph_names.end(); s++)
	    result_errh->message("%s", s->c_str());
    } catch (OpenType::Error) {
	if (errh->nerrors() == before_nerrors)
	    result_errh->message("no glyph name information");
    }
}

static void
do_tables(const OpenType::Font &otf, ErrorHandler *errh, ErrorHandler *result_errh)
{
    int before_nerrors = errh->nerrors();
    try {
	int n = otf.ntables();

	for (int i = 0; i < n; i++)
	    if (OpenType::Tag tag = otf.table_tag(i)) {
		String s = otf.table(tag);
		result_errh->message("%7u %s\n", s.length(), tag.text().c_str());
	    }
    } catch (OpenType::Error) {
	if (errh->nerrors() == before_nerrors)
	    result_errh->message("corrupted tables");
    }
}

static void
do_dump_table(const OpenType::Font &otf, OpenType::Tag tag, ErrorHandler *errh)
{
    int before_nerrors = errh->nerrors();
    try {
	if (otf.has_table(tag)) {
	    String s = otf.table(tag);
	    int written = fwrite(s.data(), 1, s.length(), stdout);
	    if (written != s.length())
		errh->error("%s", strerror(errno));
	} else
	    errh->message("no %<%s%> table", tag.text().c_str());
    } catch (OpenType::Error) {
	if (errh->nerrors() == before_nerrors)
	    errh->message("corrupted tables");
    }
}

int
main(int argc, char *argv[])
{
    Clp_Parser *clp =
	Clp_NewParser(argc, (const char * const *)argv, sizeof(options) / sizeof(options[0]), options);
    program_name = Clp_ProgramName(clp);

    ErrorHandler *errh = ErrorHandler::static_initialize(new FileErrorHandler(stderr, String(program_name) + ": "));
    Vector<const char *> input_files;
    OpenType::Tag dump_table;
    int query = 0;

    while (1) {
	int opt = Clp_Next(clp);
	switch (opt) {

	  case SCRIPT_OPT: {
	      if (!script.null())
		  usage_error(errh, "--script already specified");
	      String arg = clp->vstr;
	      int period = arg.find_left('.');
	      OpenType::Tag scr(period <= 0 ? arg : arg.substring(0, period));
	      if (scr.valid() && period > 0) {
		  OpenType::Tag lang(arg.substring(period + 1));
		  if (lang.valid()) {
		      script = scr;
		      langsys = lang;
		  } else
		      usage_error(errh, "bad language tag");
	      } else if (scr.valid())
		  script = scr;
	      else
		  usage_error(errh, "bad script tag");
	      break;
	  }

	  case QUERY_SCRIPTS_OPT:
	  case QUERY_FEATURES_OPT:
	  case QUERY_OPTICAL_SIZE_OPT:
	  case QUERY_POSTSCRIPT_NAME_OPT:
	  case QUERY_GLYPHS_OPT:
	  case QUERY_FAMILY_OPT:
	  case QUERY_FVERSION_OPT:
	  case TABLES_OPT:
	  case INFO_OPT:
	    if (query)
		usage_error(errh, "supply exactly one query type option");
	    query = opt;
	    break;

	case DUMP_TABLE_OPT:
	    if (query)
		usage_error(errh, "supply exactly one query type option");
	    if (!(dump_table = OpenType::Tag(clp->vstr)))
		usage_error(errh, "bad table name");
	    query = opt;
	    break;

	  case QUIET_OPT:
	    if (clp->negated)
		errh = ErrorHandler::default_handler();
	    else
		errh = new SilentErrorHandler;
	    break;

	  case VERBOSE_OPT:
	    verbose = !clp->negated;
	    break;

	  case VERSION_OPT:
	    printf("otfinfo (LCDF typetools) %s\n", VERSION);
	    printf("Copyright (C) 2003-2011 Eddie Kohler\n\
This is free software; see the source for copying conditions.\n\
There is NO warranty, not even for merchantability or fitness for a\n\
particular purpose.\n");
	    exit(0);
	    break;

	  case HELP_OPT:
	    usage();
	    exit(0);
	    break;

	  case Clp_NotOption:
	    input_files.push_back(clp->vstr);
	    break;

	  case Clp_Done:
	    goto done;

	  case Clp_BadOption:
	    usage_error(errh, 0);
	    break;

	  default:
	    break;

	}
    }

  done:
    if (!query)
	usage_error(errh, "supply exactly one query option");
    if (!input_files.size())
	input_files.push_back("-");
    if (script.null())
	script = Efont::OpenType::Tag("latn");

    FileErrorHandler stdout_errh(stdout);
    for (const char **input_filep = input_files.begin(); input_filep != input_files.end(); input_filep++) {
	int before_nerrors = errh->nerrors();
	String font_data = read_file(*input_filep, errh);
	if (errh->nerrors() != before_nerrors)
	    continue;

	String input_file = printable_filename(*input_filep);
	LandmarkErrorHandler cerrh(errh, input_file);
	OpenType::Font otf(font_data, &cerrh);
	if (!otf.ok())
	    continue;

	PrefixErrorHandler stdout_cerrh(&stdout_errh, input_file + ":");
	ErrorHandler *result_errh = (input_files.size() > 1 ? static_cast<ErrorHandler *>(&stdout_cerrh) : static_cast<ErrorHandler *>(&stdout_errh));
	if (query == QUERY_SCRIPTS_OPT)
	    do_query_scripts(otf, &cerrh, result_errh);
	else if (query == QUERY_FEATURES_OPT)
	    do_query_features(otf, &cerrh, result_errh);
	else if (query == QUERY_OPTICAL_SIZE_OPT)
	    do_query_optical_size(otf, &cerrh, result_errh);
	else if (query == QUERY_POSTSCRIPT_NAME_OPT)
	    do_query_postscript_name(otf, &cerrh, result_errh);
	else if (query == QUERY_GLYPHS_OPT)
	    do_query_glyphs(otf, &cerrh, result_errh);
	else if (query == QUERY_FAMILY_OPT)
	    do_query_family_name(otf, &cerrh, result_errh);
	else if (query == QUERY_FVERSION_OPT)
	    do_query_font_version(otf, &cerrh, result_errh);
	else if (query == TABLES_OPT)
	    do_tables(otf, &cerrh, result_errh);
	else if (query == DUMP_TABLE_OPT)
	    do_dump_table(otf, dump_table, &cerrh);
	else if (query == INFO_OPT)
	    do_info(otf, &cerrh, result_errh);
    }

    return (errh->nerrors() == 0 ? 0 : 1);
}
